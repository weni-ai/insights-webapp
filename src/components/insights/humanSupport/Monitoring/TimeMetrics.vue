<template>
  <section class="time-metrics">
    <p class="time-metrics__title">
      {{ $t('human_support_dashboard.time_metrics.title') }}
    </p>
    <section class="time-metrics__cards">
      <template
        v-for="(card, index) in cardDefinitions"
        :key="`${card.id}-${index}`"
      >
        <CardConversations
          class="time-metrics__card"
          :title="$t(card.titleKey)"
          :value="getCardValue(card.id)"
          :description="getCardSubValue(card.id)"
          :tooltipInfo="$t(card.tooltipKey)"
          borderRadius="full"
          activeDescriptionGap
          enableTimeIcon
          :tooltipSide="getTooltipSide(index)"
          :isLoading="isLoadingCards"
        />
      </template>
    </section>
  </section>
</template>

<script setup lang="ts">
import { onMounted, computed } from 'vue';

import CardConversations from '@/components/insights/cards/CardConversations.vue';
import { useHumanSupportMonitoring } from '@/store/modules/humanSupport/monitoring';
import { storeToRefs } from 'pinia';
import { useI18n } from 'vue-i18n';
import { formatSecondsToTime } from '@/utils/time';

type CardId =
  | 'average_time_is_waiting'
  | 'average_time_first_response'
  | 'average_time_chat';

interface CardData {
  id: CardId;
  titleKey: string;
  tooltipKey: string;
}

const cardDefinitions: CardData[] = [
  {
    id: 'average_time_is_waiting',
    titleKey: 'human_support_dashboard.time_metrics.average_time_is_waiting',
    tooltipKey:
      'human_support_dashboard.time_metrics.tooltips.average_time_is_waiting',
  },
  {
    id: 'average_time_first_response',
    titleKey:
      'human_support_dashboard.time_metrics.average_time_first_response',
    tooltipKey:
      'human_support_dashboard.time_metrics.tooltips.average_time_first_response',
  },
  {
    id: 'average_time_chat',
    titleKey: 'human_support_dashboard.time_metrics.average_time_chat',
    tooltipKey:
      'human_support_dashboard.time_metrics.tooltips.average_time_chat',
  },
];

const { t } = useI18n();

const humanSupportMonitoring = useHumanSupportMonitoring();
const { loadTimeMetricsData } = humanSupportMonitoring;
const { timeMetricsData, loadingTimeMetricsData } = storeToRefs(
  humanSupportMonitoring,
);

const isLoadingCards = computed(() => loadingTimeMetricsData.value);

const getCardValue = (id: CardId) => {
  const data = timeMetricsData.value[id];
  return formatSecondsToTime(data?.average);
};

const getCardSubValue = (id: CardId) => {
  const data = timeMetricsData.value[id];
  if (data?.max !== null && data?.max !== undefined) {
    return `${t('human_support_dashboard.time_metrics.max')}: ${formatSecondsToTime(data.max)}`;
  }
  return '';
};

const getTooltipSide = (index: number) => {
  const lastIndex = cardDefinitions.length - 1;

  if (lastIndex === index) return 'left';

  return 'top';
};

onMounted(() => {
  loadTimeMetricsData();
});
</script>

<style scoped lang="scss">
$min-height: 134px;

.time-metrics {
  display: flex;
  flex-direction: column;
  gap: $unnnic-space-3;

  &__title {
    font: $unnnic-font-display-2;
  }

  &__cards {
    display: flex;
    gap: $unnnic-space-4;
    min-height: $min-height;
  }
}
</style>
