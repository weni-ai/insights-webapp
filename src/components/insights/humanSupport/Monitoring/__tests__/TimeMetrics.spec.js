import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import TimeMetrics from '../TimeMetrics.vue';

const mockMonitoringStore = {
  loadTimeMetricsData: vi.fn(),
  timeMetricsData: {
    value: {
      average_time_is_waiting: { average: 120, max: 300 },
      average_time_first_response: { average: 45, max: 90 },
      average_time_chat: { average: 600, max: 1200 },
    },
  },
  loadingTimeMetricsData: { value: false },
};

vi.mock('@/store/modules/humanSupport/monitoring', () => ({
  useHumanSupportMonitoring: () => mockMonitoringStore,
}));

vi.mock('@/utils/time', () => ({
  formatSecondsToTime: vi.fn((seconds) => {
    if (seconds === null || seconds === undefined) return '-';
    if (seconds === 0) return '0s';
    if (seconds < 60) return `${seconds}s`;
    if (seconds < 3600)
      return `${Math.floor(seconds / 60)}m ${Math.floor(seconds % 60)
        .toString()
        .padStart(2, '0')}s`;
    return `${Math.floor(seconds / 3600)}h ${Math.floor((seconds % 3600) / 60)
      .toString()
      .padStart(2, '0')}m ${Math.floor(seconds % 60)
      .toString()
      .padStart(2, '0')}s`;
  }),
}));

vi.mock('pinia', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    storeToRefs: (store) => store,
  };
});

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: {
    en: {
      human_support_dashboard: {
        time_metrics: {
          title: 'Time Metrics',
          max: 'Max',
          average_time_is_waiting: 'Average waiting time',
          average_time_first_response: 'Average first response time',
          average_time_chat: 'Average chat duration',
          tooltips: {
            average_time_is_waiting: 'Average time users wait for support',
            average_time_first_response:
              'Average time for first agent response',
            average_time_chat: 'Average duration of chat conversations',
          },
        },
      },
    },
  },
  fallbackWarn: false,
  missingWarn: false,
});

config.global.plugins = [i18n];

describe('TimeMetrics', () => {
  let wrapper;

  const createWrapper = (storeOverrides = {}) => {
    Object.assign(mockMonitoringStore, storeOverrides);
    return mount(TimeMetrics, {
      global: {
        stubs: {
          CardConversations: true,
        },
      },
    });
  };

  const section = () => wrapper.find('[data-testid="time-metrics"]');
  const title = () => wrapper.find('[data-testid="time-metrics-title"]');
  const cards = () => wrapper.find('[data-testid="time-metrics-cards"]');

  beforeEach(() => {
    vi.clearAllMocks();
    Object.assign(mockMonitoringStore, {
      loadTimeMetricsData: vi.fn(),
      timeMetricsData: {
        value: {
          average_time_is_waiting: { average: 120, max: 300 },
          average_time_first_response: { average: 45, max: 90 },
          average_time_chat: { average: 600, max: 1200 },
        },
      },
      loadingTimeMetricsData: { value: false },
    });
    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render main section', () => {
      expect(section().exists()).toBe(true);
    });

    it('should render title', () => {
      expect(title().exists()).toBe(true);
      expect(title().text()).toBe('human_support_dashboard.time_metrics.title');
    });

    it('should render cards container', () => {
      expect(cards().exists()).toBe(true);
    });

    it('should render correct number of cards', () => {
      const cardComponents = wrapper.findAllComponents({
        name: 'CardConversations',
      });
      expect(cardComponents).toHaveLength(3);
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(section().classes()).toContain('time-metrics');
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Card definitions', () => {
    it('should have correct card definitions structure', () => {
      const vm = wrapper.vm;
      expect(vm.cardDefinitions).toHaveLength(3);

      const expectedCards = [
        {
          id: 'average_time_is_waiting',
          titleKey:
            'human_support_dashboard.time_metrics.average_time_is_waiting',
          tooltipKey:
            'human_support_dashboard.time_metrics.tooltips.average_time_is_waiting',
        },
        {
          id: 'average_time_first_response',
          titleKey:
            'human_support_dashboard.time_metrics.average_time_first_response',
          tooltipKey:
            'human_support_dashboard.time_metrics.tooltips.average_time_first_response',
        },
        {
          id: 'average_time_chat',
          titleKey: 'human_support_dashboard.time_metrics.average_time_chat',
          tooltipKey:
            'human_support_dashboard.time_metrics.tooltips.average_time_chat',
        },
      ];

      expectedCards.forEach((expected, index) => {
        expect(vm.cardDefinitions[index]).toEqual(expected);
      });
    });
  });

  describe('Data loading and store integration', () => {
    it('should call loadTimeMetricsData on mount', () => {
      expect(mockMonitoringStore.loadTimeMetricsData).toHaveBeenCalled();
    });

    it('should compute loading state correctly', () => {
      const vm = wrapper.vm;
      expect(vm.isLoadingCards).toBe(false);

      wrapper = createWrapper({ loadingTimeMetricsData: { value: true } });
      expect(wrapper.vm.isLoadingCards).toBe(true);
    });
  });

  describe('Card value formatting', () => {
    it('should format time values correctly', () => {
      const vm = wrapper.vm;

      expect(vm.getCardValue('average_time_is_waiting')).toBe('2m 00s');
      expect(vm.getCardValue('average_time_first_response')).toBe('45s');
      expect(vm.getCardValue('average_time_chat')).toBe('10m 00s');
    });

    it('should format sub values (max times) correctly', () => {
      const vm = wrapper.vm;

      expect(vm.getCardSubValue('average_time_is_waiting')).toContain(
        'Max: 5m 00s',
      );
      expect(vm.getCardSubValue('average_time_first_response')).toContain(
        'Max: 1m 30s',
      );
      expect(vm.getCardSubValue('average_time_chat')).toContain('Max: 20m 00s');
    });

    const edgeCases = [
      {
        name: 'null values',
        data: { average: null, max: null },
        expectedValue: '-',
        expectedDesc: '',
      },
      {
        name: 'undefined values',
        data: { average: undefined, max: undefined },
        expectedValue: '-',
        expectedDesc: '',
      },
      {
        name: 'zero values',
        data: { average: 0, max: 0 },
        expectedValue: '0s',
        expectedDesc: 'Max: 0s',
      },
    ];

    edgeCases.forEach(({ name, data, expectedValue, expectedDesc }) => {
      it(`should handle ${name}`, () => {
        const testData = { average_time_is_waiting: data };
        wrapper = createWrapper({
          timeMetricsData: { value: testData },
        });

        const vm = wrapper.vm;
        expect(vm.getCardValue('average_time_is_waiting')).toBe(expectedValue);

        if (expectedDesc) {
          expect(vm.getCardSubValue('average_time_is_waiting')).toContain(
            'Max:',
          );
        } else {
          expect(vm.getCardSubValue('average_time_is_waiting')).toBe('');
        }
      });
    });
  });

  describe('Tooltip side logic', () => {
    it('should return correct tooltip sides', () => {
      const vm = wrapper.vm;

      expect(vm.getTooltipSide(0)).toBe('top');
      expect(vm.getTooltipSide(1)).toBe('top');
      expect(vm.getTooltipSide(2)).toBe('left');
    });
  });

  describe('Error scenarios', () => {
    it('should handle missing store data gracefully', () => {
      wrapper = createWrapper({
        timeMetricsData: {
          value: {
            average_time_is_waiting: { average: undefined, max: undefined },
          },
        },
      });

      const vm = wrapper.vm;
      expect(vm.getCardValue('average_time_is_waiting')).toBe('-');
      expect(vm.getCardSubValue('average_time_is_waiting')).toBe('');
    });

    it('should handle malformed store data', () => {
      const malformedData = {
        average_time_is_waiting: {},
        average_time_first_response: { average: 'invalid' },
        average_time_chat: { max: 'invalid' },
      };

      wrapper = createWrapper({
        timeMetricsData: { value: malformedData },
      });

      const vm = wrapper.vm;
      expect(() => vm.getCardValue('average_time_is_waiting')).not.toThrow();
      expect(() =>
        vm.getCardSubValue('average_time_first_response'),
      ).not.toThrow();
    });
  });

  describe('CardConversations integration', () => {
    it('should pass correct props to CardConversations', () => {
      const cardComponents = wrapper.findAllComponents({
        name: 'CardConversations',
      });

      cardComponents.forEach((card) => {
        expect(card.props('borderRadius')).toBe('full');
        expect(card.props('isLoading')).toBeDefined();

        expect(card.props('activeDescriptionGap')).toBe(true);
        expect(card.props('enableTimeIcon')).toBe(true);
      });
    });

    it('should pass loading state to cards', () => {
      wrapper = createWrapper({ loadingTimeMetricsData: { value: true } });

      const cardComponents = wrapper.findAllComponents({
        name: 'CardConversations',
      });
      cardComponents.forEach((card) => {
        expect(card.props('isLoading')).toBe(true);
      });
    });
  });

  describe('Lifecycle', () => {
    it('should load time metrics data on mounted', () => {
      expect(mockMonitoringStore.loadTimeMetricsData).toHaveBeenCalled();
    });
  });

  describe('Edge cases and boundary testing', () => {
    it('should handle extremely large time values', () => {
      const extremeData = {
        average_time_is_waiting: { average: 86400, max: 172800 },
      };

      wrapper = createWrapper({
        timeMetricsData: { value: extremeData },
      });

      const vm = wrapper.vm;
      expect(() => vm.getCardValue('average_time_is_waiting')).not.toThrow();

      const result = vm.getCardValue('average_time_is_waiting');
      expect(result).toContain('h');
    });

    it('should handle negative time values gracefully', () => {
      const negativeData = {
        average_time_is_waiting: { average: -60, max: -120 },
      };

      wrapper = createWrapper({
        timeMetricsData: { value: negativeData },
      });

      const vm = wrapper.vm;
      expect(() => vm.getCardValue('average_time_is_waiting')).not.toThrow();
    });
  });

  describe('Conditional rendering', () => {
    it('should handle different store states', () => {
      const testCases = [
        {
          timeMetricsData: {
            value: {
              average_time_is_waiting: { average: 120, max: 300 },
              average_time_first_response: { average: 45, max: 90 },
              average_time_chat: { average: 600, max: 1200 },
            },
          },
          loadingTimeMetricsData: { value: false },
        },
        {
          timeMetricsData: {
            value: {
              average_time_is_waiting: { average: null, max: null },
              average_time_first_response: { average: null, max: null },
              average_time_chat: { average: null, max: null },
            },
          },
          loadingTimeMetricsData: { value: false },
        },
        {
          timeMetricsData: { value: null },
          loadingTimeMetricsData: { value: true },
        },
      ];

      const validTestCases = testCases.filter(
        (tc) => tc.timeMetricsData.value !== null,
      );

      validTestCases.forEach((testCase) => {
        wrapper = createWrapper(testCase);
        expect(section().exists()).toBe(true);
      });
    });
  });
});
