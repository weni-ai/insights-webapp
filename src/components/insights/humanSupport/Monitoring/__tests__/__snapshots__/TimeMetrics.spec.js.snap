// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimeMetrics > Component structure > should match snapshot 1`] = `
<section
  class="time-metrics"
  data-testid="time-metrics"
  data-v-3fd5bf3d=""
>
  <p
    class="time-metrics__title"
    data-testid="time-metrics-title"
    data-v-3fd5bf3d=""
  >
    human_support_dashboard.time_metrics.title
  </p>
  <section
    class="time-metrics__cards"
    data-testid="time-metrics-cards"
    data-v-3fd5bf3d=""
  >
    
    <card-conversations-stub
      activedescriptiongap="true"
      borderradius="full"
      class="time-metrics__card"
      data-v-3fd5bf3d=""
      description="Max: 5m 00s"
      enabletimeicon="true"
      isclickable="false"
      isloading="false"
      title="human_support_dashboard.time_metrics.average_time_is_waiting"
      tooltipinfo="human_support_dashboard.time_metrics.tooltips.average_time_is_waiting"
      tooltipside="top"
      value="2m 00s"
    />
    <card-conversations-stub
      activedescriptiongap="true"
      borderradius="full"
      class="time-metrics__card"
      data-v-3fd5bf3d=""
      description="Max: 1m 30s"
      enabletimeicon="true"
      isclickable="false"
      isloading="false"
      title="human_support_dashboard.time_metrics.average_time_first_response"
      tooltipinfo="human_support_dashboard.time_metrics.tooltips.average_time_first_response"
      tooltipside="top"
      value="45s"
    />
    <card-conversations-stub
      activedescriptiongap="true"
      borderradius="full"
      class="time-metrics__card"
      data-v-3fd5bf3d=""
      description="Max: 20m 00s"
      enabletimeicon="true"
      isclickable="false"
      isloading="false"
      title="human_support_dashboard.time_metrics.average_time_chat"
      tooltipinfo="human_support_dashboard.time_metrics.tooltips.average_time_chat"
      tooltipside="left"
      value="10m 00s"
    />
    
  </section>
</section>
`;
